webapi/src/main/java/org/sakaiproject/webapi/controllers/CalendarController.java

```diff
@@ -97,40 +100,76 @@
+            .stream()
+            .filter(s -> s.getCreatedDate().toInstant().isAfter(projectCutoff))
+            .map(s -> s.getId())
+            .collect(Collectors.toList());
+
+        Map<String, String> propCrit = new HashMap<>();
+
+        // Now grab all ids of sites in the current academic sessions
+        courseManagementService.getCurrentAcademicSessions().forEach(as -> {
+
+            propCrit.put(Site.PROP_SITE_TERM, as.getTitle());
+            allSites.addAll(siteService.getSiteIds(SiteService.SelectionType.ACCESS, "course", null, propCrit, null, null));
+        });
+
+        Map<String, Object> data = new HashMap<>();
+        data.put("events", allSites.stream().map(s -> getEventsForSite(s))
+            .flatMap(Collection::stream).collect(Collectors.toList()));
+        data.put("days", serverConfigurationService.getInt("webapi.calendar.days", DEFAULT_DAYS));
+        return data;
 	}
 
 	@GetMapping(value = "/sites/{siteId}/calendar", produces = MediaType.APPLICATION_JSON_VALUE)
-    public List<CalendarEventRestBean> getSiteCalendar(@PathVariable String siteId) throws UserNotDefinedException {
+    public Map<String, Object> getSiteCalendar(@PathVariable String siteId) throws UserNotDefinedException {
+
+		checkSakaiSession();
+
+        //try {
+            Map<String, Object> data = new HashMap<>();
+            data.put("events", getEventsForSite(siteId));
+            data.put("days", serverConfigurationService.getInt("webapi.calendar.days", DEFAULT_DAYS));
+            return data;
+        /*} catch (Exception e) {
+            log.error("Failed to get calendar events for site {}. An empty map will be generated : {}", siteId, e.toString());
+            return Collections.emptyMap();
+        }*/
+	}
+
+    private List<CalendarEventRestBean> getEventsForSite(String siteId) {
 
-		Session session = checkSakaiSession();
         try {
             return ((List<CalendarEvent>) calendarService.getCalendar(calendarService.calendarReference(siteId, "main"))
-                .getEvents(null, null))
+                .getEvents(getRange(), null))
                 .stream().map(convert).collect(Collectors.toList());
+        } catch (IdUnusedException idue) {
+            // Normal. If the calendar hasn't been added to a site, this is what we'll get.
         } catch (Exception e) {
-            log.error("Failed to get calendar events for site {}. An empty list will be generated", siteId);
-            return Collections.<CalendarEventRestBean>emptyList();
+            log.error("Failed to get calendar events for site {}. An empty map will be generated : {}", siteId, e.toString());
         }
-	}
+
+        return Collections.<CalendarEventRestBean>emptyList();
+    }
+
+    private TimeRange getRange() {
+
+        Instant now = Instant.now();
+        Instant inAYear = now.plus(serverConfigurationService.getInt("webapi.calendar.days", DEFAULT_DAYS), ChronoUnit.DAYS);
```

Comment:
This method is potentially called many times inside a loop, I'm not very keen on loading the same property over and over again.

Reply:
Good spot, thanks.